mindmeld/components/custom_action.py

```diff
@@ -0,0 +1,187 @@
+        }
+        responder_json = DialogueResponder.to_json(responder)
+        return {
+            "request": request_json,
+            "responder": {field: responder_json[field] for field in RESPONSE_FIELDS},
+            "action": self._name,
+        }
+
+    def invoke(self, request, responder, async_mode=False):
+        """Invoke the custom action with Request and Responder and return True if the action is
+        executed successfully, False otherwise. Upon successful execution, we update the Frame
+        and Directives of the Responder object.
+
+        Args:
+            request (Request)
+            responder (DialogueResponder)
+            async_mode (bool)
+
+        Returns:
+            (bool)
+        """
+        if not self.url:
+            raise CustomActionException(
+                "No URL is given for custom action {}.".format(self._name)
+            )
+
+        json_data = self.get_json_payload(request, responder)
+
+        try:
+            if async_mode:
+                # returning the coroutine to be awaited elsewhere
+                return self._process_async(json_data, responder)
+            else:
+                return self._process(json_data, responder)
+        except ConnectionError:
+            logger.error(
+                "Connection error trying to reach custom action server %s.", self.url
+            )
+            return False
+
+    async def invoke_async(self, request, responder):
+        """Asynchronously invoke the custom action with Request and Responder and return True if
+        the action is executed successfully, False otherwise. Upon successful execution, we update
+        the Frame and Directives of the Responder object.
+
+        Args:
+            request (Request)
+            responder (DialogueResponder)
+
+        Returns:
+            (bool)
+        """
+        return await self.invoke(request, responder, async_mode=True)
+
+    def _process(self, json_data, responder):
+        status_code, result_json = self.post(json_data)
+        return self._process_post_response(status_code, result_json, responder)
+
+    async def _process_async(self, json_data, responder):
+        status_code, result_json = await self.post_async(json_data)
+        return self._process_post_response(status_code, result_json, responder)
+
+    def _process_post_response(self, status_code, result_json, responder):
+        if status_code == 200:
+            for field in RESPONSE_FIELDS:
+                if field not in result_json:
+                    logger.warning(
+                        "`%s` not in the response of custom action %s",
+                        field,
+                        self._name,
+                    )
+            if self.overwrite:
+                responder.frame = result_json.get("frame", {})
+                responder.directives = result_json.get("directives", [])
+            else:
+                responder.frame.update(result_json.get("frame", {}))
```

Comment:
update overwrites the existing key. I dont think this is the intention when supplying overwrite=False.

Reply:
There is no population of the params field here.

Reply:
We assume that if the user changes the key value in the frame that we send them, they intend for that change to effect. The `overwrite` field distinguishes the case where the user wants the returned frame to be a "hard" change (as it is) or a soft change (merging and updating with any existing key/value).

Reply:
What about the common case where the user wants to only update keys that he/she changed, and not any other keys? We could accomodate this by doing the following:
1. If overwrite is True and user provides keys to change, we overwrite those keys that the user wants to change. If they want to overwrite ALL the keys (as you mentioned), they can also do that by supplying all the keys to overwrite.
2. If overwrite is False and user provides keys whose values collide with existing keys, those collision keys are NOT updated.

I believe the above behavior is the usual definition of what is means to "override" a key or not.